skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -0,0 +1,430 @@
+
+    neighborhood[24] = img[p+1, r-1, c+1]
+    neighborhood[25] = img[p+1, r,   c+1]
+    neighborhood[26] = img[p+1, r+1, c+1]
+
+
+###### look-up tables
+def fill_Euler_LUT():
+    """ Look-up table for preserving Euler characteristic.
+
+    This is column $\delta G_{26}$ of Table 2 of [Lee94]_.
+    """
+    cdef int arr[128]
+    arr[:] = [1, -1, -1, 1, -3, -1, -1, 1, -1, 1, 1, -1, 3, 1, 1, -1, -3, -1,
+                 3, 1, 1, -1, 3, 1, -1, 1, 1, -1, 3, 1, 1, -1, -3, 3, -1, 1, 1,
+                 3, -1, 1, -1, 1, 1, -1, 3, 1, 1, -1, 1, 3, 3, 1, 5, 3, 3, 1,
+                 -1, 1, 1, -1, 3, 1, 1, -1, -7, -1, -1, 1, -3, -1, -1, 1, -1,
+                 1, 1, -1, 3, 1, 1, -1, -3, -1, 3, 1, 1, -1, 3, 1, -1, 1, 1,
+                 -1, 3, 1, 1, -1, -3, 3, -1, 1, 1, 3, -1, 1, -1, 1, 1, -1, 3,
+                 1, 1, -1, 1, 3, 3, 1, 5, 3, 3, 1, -1, 1, 1, -1, 3, 1, 1, -1]
+    cdef ndarray LUT = np.zeros(256, dtype=np.intc)
+    LUT[1::2] = arr
+    return LUT
+cdef int[::1] LUT = fill_Euler_LUT()
+
+
+# Fill the look-up table for indexing octants for computing the Euler
+# characteristic. See is_Euler_invariant routine below.
+{{py:
+_neighb_idx = [[2, 1, 11, 10, 5, 4, 14],      # NEB
+               [0, 9, 3, 12, 1, 10, 4],       # NWB
+               [8, 7, 17, 16, 5, 4, 14],      # SEB
+               [6, 15, 7, 16, 3, 12, 4],      # SWB
+               [20, 23, 19, 22, 11, 14, 10],  # NEU
+               [18, 21, 9, 12, 19, 22, 10],   # NWU
+               [26, 23, 17, 14, 25, 22, 16],  # SEU
+               [24, 25, 15, 16, 21, 22, 12],  # SWU
+              ]
+}}
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef bint is_Euler_invariant(pixel_type neighbors[],
+                             int[::1] lut):
+    """Check if a point is Euler invariant.
+
+    Calculate Euler characteristic for each octant and sum up.
+
+    Parameters
+    ----------
+    neighbors
+        neighbors of a point
+    lut
+        The look-up table for preserving the Euler characteristic.
+
+    Returns
+    -------
+    bool (C bool, that is)
+
+    """
+    cdef int n, euler_char = 0
+    {{for _octant in range(8)}}
+
+    # octant {{_octant}}:
+    n = 1
+    {{for _j in range(7):}}
+    {{py: _idx = _neighb_idx[_octant][_j]}}
+    if neighbors[{{_idx}}] == 1:
+        n |= {{1 << (7 - _j)}}
+
+    {{endfor}}
+    euler_char += lut[n]
+    {{endfor}}
+    return euler_char == 0
+
+
+cdef inline bint is_endpoint(pixel_type neighbors[]):
+    """An endpoint has exactly one neighbor in the 26-neighborhood.
+    """
+    # The center pixel is counted, thus r.h.s. is 2
+    cdef int s = 0, j
+    for j in range(27):
+        s += neighbors[j]
+    return s == 2
+
+
+cdef bint is_simple_point(pixel_type neighbors[]):
+    """Check is a point is a Simple Point.
+
+    This method is named "N(v)_labeling" in [Lee94]_.
+    Outputs the number of connected objects in a neighborhood of a point
+    after this point would have been removed.
```

Comment:
Um, no, that's not what it outputs...? It would be useful to add the *definition* of a simple point to this docstring.